AK/IntegralMath.h

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2022, Leon Albrecht <leon2002.la@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/BuiltinWrappers.h>
+#include <AK/Concepts.h>
+#include <AK/Types.h>
+
+#include <AK/Format.h>
+
+namespace AK {
+
+template<Integral T>
+constexpr T exp2(T exponent)
+{
+    return 1u << exponent;
+}
+
+template<Integral T>
+constexpr T log2(T x)
+{
+    return x ? (8 * sizeof(T) - 1) - count_leading_zeroes(static_cast<MakeUnsigned<T>>(x)) : 0;
+}
+
+template<Integral I>
+inline I pow(I base, I exponent)
+{
+    // https://en.wikipedia.org/wiki/Exponentiation_by_squaring
+    if (exponent < 0)
+        return 0;
+    if (exponent == 0)
+        return 1;
+
+    I res = 1;
+    while (exponent > 1) {
+        if (exponent & 1)
+            res *= base;
+        base *= base;
+        exponent /= 2u;
+    }
+    return res * base;
```

Comment:
We can avoid the extra `exponent == 0` branch if we tweak the loop condition a bit:

```suggestion
    if (exponent < 0)
        return 0;

    I res = 1;
    while (exponent > 0) {
        if (exponent & 1)
            res *= base;
        base *= base;
        exponent /= 2u;
    }
    return res;
```